### PR TITLE
Support ML Inference Search Processor Writing to Search Extension 

### PR DESCRIPTION
**Pull request by mingshl**
### Description
Previously, ml inference search processor only writing prediction results to the hits. Now, Support ML Inference Search Processor Writing to Search Extension when many to one inference. 

Note that it's not supported for one to one inference because the order of one to one inference matters and other processors might rerank the order and mess up the order to match one to one model input and prediction results 

### Related Issues
https://github.com/opensearch-project/ml-commons/issues/2878

### Sample Test Case

```

PUT /review_string_index/_doc/1
{
  "review": "Dr. Eric Goldberg is a fantastic doctor who has correctly diagnosed every issue that my wife and I have had. Unlike many of my past doctors, Dr. Goldberg is very accessible and we have been able to schedule appointments with him and his staff very quickly. We are happy to have him in the neighborhood and look forward to being his patients for many years to come." ,
  "label":"5 stars"
}

PUT /review_string_index/_doc/2
{
  "review": "happy visit" ,
  "label":"5 stars"
}


PUT /review_string_index/_doc/3
{
  "review": "sad place" ,
  "label":"1 stars"
}

PUT /_search/pipeline/my_pipeline_request_review_llm
{
  "response_processors": [
    {
      "ml_inference": {
        "tag": "ml_inference",
        "description": "This processor is going to run llm",
        "model_id": "uhkETJIB5-xYSMo_SPet",
        "function_name": "REMOTE",
        "input_map": [
          {
            "context": "review"
          }
        ],
        "output_map": [
          {
            "ext.ml_inference.params.model_response": "response" 
          }
        ],
        "model_config": {
          "prompt":"\n\nHuman: You are a professional data analysist. You will always answer question based on the given context first. If the answer is not directly shown in the context, you will analyze the data and find the answer. If you don't know the answer, just say I don't know. Context: ${parameters.context.toString()}. \n\n Human: please summarize the documents \n\n Assistant:"
        },
        "ignore_missing": false,
        "ignore_failure": false,
        "one_to_one":false
      }
    }
  ]
}

GET /review_string_index/_search?search_pipeline=my_pipeline_request_review_llm
{"query":{
  "match_all": {}
}
}
```
returnning 

```
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 3,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "review_string_index",
        "_id": "1",
        "_score": 1,
        "_source": {
          "review": "Dr. Eric Goldberg is a fantastic doctor who has correctly diagnosed every issue that my wife and I have had. Unlike many of my past doctors, Dr. Goldberg is very accessible and we have been able to schedule appointments with him and his staff very quickly. We are happy to have him in the neighborhood and look forward to being his patients for many years to come.",
          "label": "5 stars"
        }
      },
      {
        "_index": "review_string_index",
        "_id": "2",
        "_score": 1,
        "_source": {
          "review": "happy visit",
          "label": "5 stars"
        }
      },
      {
        "_index": "review_string_index",
        "_id": "3",
        "_score": 1,
        "_source": {
          "review": "sad place",
          "label": "1 stars"
        }
      }
    ]
  },
  "ext": {
    "ml_inference": {
      "llm_response": """ Based on the context provided:

- The first document is a positive review of Dr. Eric Goldberg from a patient. It praises Dr. Goldberg for correctly diagnosing issues for the patient and their wife. It also notes that Dr. Goldberg is very accessible and appointments can be scheduled quickly with him and his staff. The patient expresses happiness that Dr. Goldberg is in their neighborhood and looks forward to being his patient for many years.

- The second document just says "happy visit". 

- The third document says "sad place".

- In summary, the first document positively reviews a doctor, Dr. Eric Goldberg. The other two documents don't provide much context on their own, just mentioning a "happy visit" and "sad place"."""
    }
  }
}
```

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ohltyler** [comment]
Does this change the default functionality of many-to-one, such that the outputs in the output map will always be placed outside of the individual document `_source`s, or is that still an option? I don't know if there's any good use cases to support the latter.

**Comment by mingshl** [comment]
> Does this change the default functionality of many-to-one, such that the outputs in the output map will always be placed outside of the individual document `_source`s, or is that still an option? I don't know if there's any good use cases to support the latter.

The output mapping will allow users to define whether they want to save to extension or the document source. If the output mappings points to prefix `ext.ml_inference`, the model output will save to extension, else other mappings will default saving to document source

**Comment by ohltyler** [comment]
> > Does this change the default functionality of many-to-one, such that the outputs in the output map will always be placed outside of the individual document `_source`s, or is that still an option? I don't know if there's any good use cases to support the latter.
> 
> The output mapping will allow users to define whether they want to save to extension or the document source. If the output mappings points to prefix `ext.ml_inference`, the model output will save to extension, else other mappings will default saving to document source

Got it, thanks.

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 1]
```diff
@@ -774,6 +802,21 @@ public MLInferenceSearchResponseProcessor create(
...
+                    for (String key : outputMap.keySet()) {
+                        if (key.startsWith(EXTENSION_PREFIX)) {
+                            writeToSearchExtension = true;
+                            break;
+                        }
+                    }
+                }
+            }
+            if (writeToSearchExtension & oneToOne) {
+                throw new IllegalArgumentException("Writing model response to search extension does not support when one_to_one is true.");
```
Comment (brianf-aws) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:818]: What do you feel about this?
```
Writing model response to search extension is not supported when one_to_one is true
```
or
```
Write model response to search extension is not supported when one_to_one is true
```

File: plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java [hunk 2]
```diff
@@ -978,14 +1055,18 @@ public void testProcessResponseCreateRewriteResponseListenerExceptionIgnoreFailu
...
         RuntimeException mockException = new RuntimeException("Mock exception");
         AtomicInteger callCount = new AtomicInteger(0);
-        ;
+
         when(mockResponse.getHits()).thenAnswer(invocation -> {
 
             int count = callCount.getAndIncrement();
 
-            if (count == 2) {
+            if (count == 6) {
```
Comment (brianf-aws) [inline plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java:1069]: Whats the significance of the number 6? If its tied to a object with a size perhaps we can use that instead? also maybe we can add a inequality since strict equality, I know you are using an AtomicInteger but the paranoid side of me feels that way 😅

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 3]
```diff
@@ -545,22 +558,37 @@ public void onResponse(Map<Integer, MLOutput> multipleMLOutputs) {
...
+                                                    sourceAsMapWithInference.remove(newDocumentFieldName);
+                                                    sourceAsMapWithInference.put(newDocumentFieldName, modelOutputValuePerDoc);
+                                                } else {
+                                                    logger
+                                                        .debug(
+                                                            "{} already exists in the search response hit. Skip processing this field.",
+                                                            newDocumentFieldName
+                                                        );
+                                                    // TODO when the response has the same field name, should it throw exception? currently,
+                                                    // ingest processor quietly skip it
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:587]: What is this toDo? Seems like this is a ToDO for ingest processor but listed here in the search processor?

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 4]
```diff
@@ -774,6 +802,21 @@ public MLInferenceSearchResponseProcessor create(
...
+            if (outputMaps != null) {
+                for (Map<String, String> outputMap : outputMaps) {
+                    for (String key : outputMap.keySet()) {
+                        if (key.startsWith(EXTENSION_PREFIX)) {
+                            writeToSearchExtension = true;
+                            break;
+                        }
+                    }
+                }
+            }
```
Comment (Zhangxunmt) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:816]: if (outputMaps != null && outputMaps.stream().anyMatch(
        outputMap -> outputMap.keySet().stream().anyMatch(key -> key.startsWith(EXTENSION_PREFIX))
    )) {
    writeToSearchExtension = true;
}
Please check using this Steam to avoid 4 or 5 nested layers.

**Review by Zhangxunmt — DISMISSED** [review]

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 3 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:587]: It's the same logic in ingest and search, currently it will quietly skip adding to response/documents when there is the same name field name already existed in the response/documents.

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 4 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:816]: good call out, I will make this change

File: plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java (hunk 2 above)
Comment (mingshl) [reply plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java:1069]: it was the 2nd time that it calls mockResponse.getHits() then throw exceptions, but now because I made the changes that calls 4 more times earlier, so it's calling the 6 time now to throw the exception at the tested function.

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 1 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:818]: will take the second one, it seems more formal this way

**Review by Zhangxunmt — DISMISSED** [review]

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 5]
```diff
@@ -774,6 +802,18 @@ public MLInferenceSearchResponseProcessor create(
                         + ". Please adjust mappings."
                 );
             }
+            boolean writeToSearchExtension = false;
+
+            if (outputMaps != null
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:807]: Reduce nested streams: Instead of nesting two stream() operations, Seems like we can use flatMap() to transform each outputMap into a stream of keys directly.

Filter nulls: Added a null check (filter(Objects::nonNull)) to prevent NullPointerException in case outputMaps or keySet() contain null elements.


```
if (outputMaps != null) {
    writeToSearchExtension = outputMaps.stream()
        .filter(Objects::nonNull) // To avoid potential NullPointerExceptions from null outputMaps
        .flatMap(outputMap -> outputMap.keySet().stream())
        .anyMatch(key -> key.startsWith(EXTENSION_PREFIX));
}
```

File: plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java [hunk 6]
```diff
@@ -503,6 +505,81 @@ public void onFailure(Exception e) {
         verify(client, times(1)).execute(any(), any(), any());
     }
 
+    /**
+     * Tests the successful processing of a response with a single pair of input and output mappings.
+     * read the query text into model config
+     * with query extensions
+     * @throws Exception if an error occurs during the test
+     */
+    public void testProcessResponseSuccessWriteToExt() throws Exception {
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java:514]: let's add `@Test` for every test methods.

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 7]
```diff
@@ -774,6 +802,21 @@ public MLInferenceSearchResponseProcessor create(
                         + ". Please adjust mappings."
                 );
             }
+            boolean writeToSearchExtension = false;
```
[the remark this answers is not in the document]
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:805]: isn't this one false?

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 8]
```diff
@@ -84,6 +84,7 @@ public class MLInferenceSearchResponseProcessor extends AbstractProcessor implem
     // it can be overwritten using max_prediction_tasks when creating processor
     public static final int DEFAULT_MAX_PREDICTION_TASKS = 10;
     public static final String DEFAULT_OUTPUT_FIELD_NAME = "inference_results";
+    public static final String EXTENSION_PREFIX = "ext.ml_inference";
```
[the remark this answers is not in the document]
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:87]: added

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:805]: I meant assigning false is unnecessary which is already default value.

[the remark this answers is not in the document]
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:805]: oh if removing assigning false, I will get `Variable 'writeToSearchExtension' might not have been initialized`

[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:805]: oh ok, ignore my comment then.

File: plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java (hunk 6 above)
Comment (mingshl) [reply plugin/src/test/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessorTests.java:514]: added

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.action.search.SearchResponseSections;
+import org.opensearch.action.search.ShardSearchFailure;
+import org.opensearch.core.xcontent.XContentBuilder;
+
+public class MLInferenceSearchResponse extends SearchResponse {
```
Comment (saratvemulapalli) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java:15]: Not related to the PR, just curious why ML Inference processor is housed in the plugin directory and not in search-processor directory ?

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 5 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java]: even better!

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public Map<String, Object> getParams() {
+        return this.params;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        innerToXContent(builder, params);
```
[the remark this answers is not in the document]
Comment (brianf-aws) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java:46]: Hey Dhrubo, From my experience when you are trying to get the `_source` Map<String, Object> object into the format the search response has you need to to convert it to xContent for it to go back. Here is a snippet of program I did

```
                XContentBuilder builder = XContentBuilder.builder(mediaTypeAndSourceMapTuple.v1().xContent());
                builder.map(sourceAsMap);
                hit.sourceRef(BytesReference.bytes(builder));
```
My hunch is that since this new type of search response is doing a different interpretation of the object it will invoke this method when it converts it back.
I say this because of this line ` MLInferenceSearchResponse extends SearchResponse` and if you check the processor code it creates a `MLInferenceSearchResponse` not the usual `SearchResponse`

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java (hunk 9 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java:15]: we had this discussion earlier, the main reason to host the ml inference processor in plugin folder is to get closers to the model, so that it's easier to get local model and remote model and compose their model inputs,

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        this.params = params;
+    }
+
+    public void setParams(Map<String, Object> params) {
+        this.params = params;
+    }
+
+    public Map<String, Object> getParams() {
+        return this.params;
+    }
```
Comment (brianf-aws) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java:41]: Maybe we can use lombok 

```
    @Getter
    @Setter
    private Map<String, Object> params;
```

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 12]
```diff
@@ -14,13 +14,7 @@
...
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:17]: Avoid using import *

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java [hunk 13]
```diff
@@ -158,7 +155,19 @@ public void processResponseAsync(
 
             // if many to one, run rewriteResponseDocuments
             if (!oneToOne) {
-                rewriteResponseDocuments(response, responseListener);
+                // use MLInferenceSearchResponseProcessor to allow writing to extension
+                MLInferenceSearchResponse mLInferenceSearchResponse = new MLInferenceSearchResponse(
+                    null,
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:160]: In your SearchResponse, I noticed:

```
        if (this.params != null) {
            builder.startObject(EXT_SECTION_NAME);
            builder.field(MLInferenceSearchResponseProcessor.TYPE, this.params);

            builder.endObject();
        }
        ```
        
        But we are sending null from here?

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 12 above)
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:17]: You can just add this in your intellij settings: https://stackoverflow.com/questions/26225989/how-do-i-stop-wildcard-java-package-imports-in-intellij

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java (hunk 11 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponse.java:41]: there is only one attribute in the class, so it's ok to keep the short methods for readability. But if we add more attributes to this class later, it's worth to use lombok to reduce the codes.

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 13 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java:160]: if a search response doesn't have a search extension, we don't add null params to the search extension,

for example, a search response with null params looks like this: 

```
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 3,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "review_string_index",
        "_id": "1",
        "_score": 1,
        "_source": {
          "review": "Dr. Eric Goldberg is a fantastic doctor who has correctly diagnosed every issue that my wife and I have had. Unlike many of my past doctors, Dr. Goldberg is very accessible and we have been able to schedule appointments with him and his staff very quickly. We are happy to have him in the neighborhood and look forward to being his patients for many years to come.",
          "label": "5 stars"
        }
      }
}
```

for example, a search response with some params looks like this: 
```
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 3,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "review_string_index",
        "_id": "1",
        "_score": 1,
        "_source": {
          "review": "Dr. Eric Goldberg is a fantastic doctor who has correctly diagnosed every issue that my wife and I have had. Unlike many of my past doctors, Dr. Goldberg is very accessible and we have been able to schedule appointments with him and his staff very quickly. We are happy to have him in the neighborhood and look forward to being his patients for many years to come.",
          "label": "5 stars"
        }
      }
  },
  "ext": {
    "ml_inference": {
      "llm_response": """ Based on the context provided:

- The first document is a positive review of Dr. Eric Goldberg from a patient. It praises Dr. Goldberg for correctly diagnosing issues for the patient and their wife. It also notes that Dr. Goldberg is very accessible and appointments can be scheduled quickly with him and his staff. The patient expresses happiness that Dr. Goldberg is in their neighborhood and looks forward to being his patient for many years.

- The second document just says "happy visit". 

- The third document says "sad place".

- In summary, the first document positively reviews a doctor, Dr. Eric Goldberg. The other two documents don't provide much context on their own, just mentioning a "happy visit" and "sad place"."""
    }
  }
}
```

Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java]: following up in this discussion, I added the check to see if the processing response is in the type of  MLInferenceSearchResponse, initiate it when it's not a  MLInferenceSearchResponse . https://github.com/opensearch-project/ml-commons/pull/3061/commits/f7682e0f60212d31e6ae6d8e49f265a965459640

File: plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java (hunk 12 above)
Comment (mingshl) [reply plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java]: reverted  import *

**Review by dhrubo-os — APPROVED** [review]

**Review by Zhangxunmt — APPROVED** [review]
